core/src/client/async_client/mod.rs

```diff
@@ -174,22 +174,19 @@ impl ClientBuilder {
 		let ping_interval = self.ping_interval;
 		let (on_exit_tx, on_exit_rx) = oneshot::channel();
 
-		tokio::spawn(async move {
-			background_task(
-				sender,
-				receiver,
-				from_front,
-				err_tx,
-				max_buffer_capacity_per_subscription,
-				ping_interval,
-				on_exit_rx,
-			)
-			.await;
-		});
+		// NOTE: this is a an unbounded channel to ensure that sends never blocks
+		// if the send task is got backpressured or something else.
+		let (msg_tx, msg_rx) = mpsc::unbounded_channel();
+		let (close_tx, close_rx) = mpsc::channel(1);
+
+		tokio::spawn(send_task(sender, msg_rx, ping_interval, close_tx.clone()));
+		tokio::spawn(read_task(receiver, msg_tx, from_front, max_buffer_capacity_per_subscription, close_tx));
```

Comment:
To check my understanding, it looks like:

- `read_task` reads messages from the frontend _or_ the backend. If they come from the frontend, the messagesa re passed on to `send_task` via an unbounded channel
- `send_task` receives messages from `read_task` via the unbounded channel and sends them to the backend.

So, if I make lots of `request`'s with the client, and the backend isn't ready to receive them yet, then they will just accumulate in the unbounded channel. If we were to use a bounded channel there, then the `read_task` would block because it's the one sending messages to this channel, and we'd end up at square one again where we'd deadlock on reading messages from the backend because `read_task` is blocked trying to send something.

So in this PR it looks like:

```
client ----> read_task <---- backend
                 +
              manager
                 |
                 | (unbounded channel)
                 v
             send_task ----> backend
``` 

Ie client sends request to `read_task`, which stores it in manager to wait for a response, and sends it on to the backend via `send_task`.
When response comes back to `read_task`, manager checked and value given back to user via oneshot.

Here, I think backpressure is lost in the client because an unlimited amount of `requests` can be made and the client will take them all happily and put them into the unbounded channel.

I think the earlier iteration of this with the mutex looked more like this:

```
client ----> from_front ----> backend
                 |
                 v
client <----- manager
                 ^
                 |
             from_back <---- backend
```

In this, both sides have to briefly lock the manager to insert requests or remove requests to respond to, but `from_front` and `from_back` can never block each other; if one is stuck, the other will always try to keep working, and backpressure will go back to the client calling `request` or whatever as needed.

So basically, I think what you did earlier is the "cleaner" approach, and avoids coupling the two tasks (or having an unbounded channel)

Reply:
(and if the mutexed request manager was to become a bottleneck, it should be quite easy to shard the map (ie use soemthing like `dashmap`) so that accesses don't block eachother on each side so much :))

Reply:
I think your understanding of the code is correct however the old code had still had a coupling between the send/receive task because if the receive task couldn't propagate the subscription to the user i.e, if the subscription was dropped.

Then the client would an unsubscribe call to the server, it was done here: https://github.com/paritytech/jsonrpsee/blob/8478e142d557543627f5cab38101b23c51f1233c/core/src/client/async_client/mod.rs#L861-#L863

But yeah that would be much more unlikely to occur because after the refactoring all goes via that channel...

Reply:
> however the old code had still had a coupling between the send/receive task because if the receive task couldn't propagate the subscription to the user i.e, if the subscription was dropped.

Aah I see, so in the mutex version the receive task would still try sending things to the send task.

I guess I'm wondering whether it makes sense to havea  flow more like what I assumed it was anyway:

```
client ----> from_front ----> backend
                 |
                 v
client <----- manager
                 ^
                 |
             from_back <---- backend
```

That way, there is for sure no way that one of the tasks can ever block the other one from proceeding (at most they will lock the manager very briefly when they need to update things in it).

But maybe I'm missing something :)

Reply:
I have reverted the latest stuff but still it's not as clean as having everything in the manager. Still the receive task can send unsubscribe messages via the front/back channel but it's not an unbounded channel at least.

In order to have something as clean as you suggested we would need to have shared queue in the RequestManager where stuff can inserted and pop:ed from to send back unsubscribe calls.

It's quite to hard to do unless we split up the RequestManager to use several locks for the data and that will become even harder to maintain I think.....